### ISSUE TITLE
Support polygon fills

### ISSUE THREAD

**Issue by pagameba** [openlayers/ol3#2208]
It might be nice to support polygon fills, which can be accomplished with the [canvas `createPattern`](http://www.html5canvastutorials.com/tutorials/html5-canvas-patterns-tutorial/) method.

**Comment by MrPl0p** [comment]
Hello, 

What do you think of this ?

![image](https://cloud.githubusercontent.com/assets/11309450/6483923/dc8a8922-c276-11e4-859c-ea03e1a80fed.png)

Thank you

**Comment by ahocevar** [comment]
Looks promising!

**Comment by TheHeat** [comment]
Hi, I'd _love_ this feature. Looking good @MrPl0p

**Comment by tremby** [comment]
Yes please.

**Comment by MrPl0p** [comment]
Hello guys, 

Here's a white paper I wrote during my internship. 
http://www.quadratic.be/download/Advanced_styles_in_OpenLayers_3.pdf
I talk about OL3 styles research and development.
I hope it will be useful for the OL3 community, anyway I enjoyed to do it.

**Comment by bartvde** [comment]
thanks for sharing @MrPl0p

**Comment by BevolutionZ** [comment]
thanks for the document !

**Comment by effetx** [comment]
Many many thanks @MrPl0p for the document. I read with attention what you wrote but I yet have some problems to "dress" the vector layes

  var image_repeat= new Image();
  image_repeat.src="graphicfill.png";

  var vectorlayer= new ol.layer.Vector({source: new ol.source.GeoJSON({projection: 'EPSG:3857',url: 'file.ge'}),opacity: 0.8,              
              style: (function() {          
              return function(feature, resolution) {
              var ruleStyle1 = [new ol.style.Style({fill: 
                new ol.style.Fill({
                  image:[image_repeat,0.4]
                }),
                stroke: new ol.style.Stroke({color:'#bbb',width:3})          
          })];
      
                        var attributo='singolo';
          if ((attributo === '')||(attributo == 'singolo')) {
            return ruleStyle1;}
            };
                            })()});
  map.addLayer(vectorlayer);
  
  
I hack the ol-debug.js file with the code proposed in your document (or I hope to having done this, but probability I make some mistakes) adding a row in ol.style.fill function and adding prototype.getImage function, in this way:

ol.style.Fill = function(opt_options) {
  var options = goog.isDef(opt_options) ? opt_options : {};
  this.color_ = goog.isDef(options.color) ? options.color : null;
  this.image_=goog.isDef(options.image) ? options.image : null;
  this.checksum_ = undefined;
};

ol.style.Fill.prototype.getColor = function() {
  return this.color_;
};

ol.style.FillImage.prototype.getImage= function(){
  return this._image;
}

But the result is always the vector layer with black color. Can you help me?

**Comment by MrPl0p** [comment]
Hi @effetx , 

I'm currently completing the end of my studies with a huge work which doesn't let me much time sorry. But as soon as it finish, I will answer to your question with more details. Promissed.

Thanks for your interest.

**Comment by MrPl0p** [comment]
Hi @effetx,

I think you forgot to hook the ol.render.canvas.Replay.prototype.replay_ function() as describe in my document.
If you want to use an image instead of a fill color, you have to alter this function to assigne the canvas pattern used for you image to fillstyle context at the SET_FILL_STYLE insctruction. Look up at the page 17 of the document.

If you want, I could make you available an extension of the OL3 librairy which contains all the overrided code.

Bye.

**Comment by effetx** [comment]
Hi MrPl0p,
thank you very much for your answer. Actually I hook the function following the instructions in your document but without results: all my polygons are completely black.

This is the hook that i used in ol-debug.js:

```
case ol.render.canvas.Instruction.SET_FILL_STYLE:
        if (goog.isString(instruction[1])) {
            context.fillStyle = instruction[1];
        }
        if (goog.isArray(instruction[1])) {
            if (!(instruction[1][0] instanceof(Image))) {
                context.fillStyle = 'black';
            }
            else {
                    if (instruction[1].length==2) {
                      setImageAlpha(instruction[1][0], instruction[1][1]);
                    }
                    var pattern=context.createPattern(instruction[1][0],'repeat');
                    context.fillStyle = pattern;
                  }
        }
```


and this is the original version:
```
goog.asserts.assert(goog.isString(instruction[1]),
            '2nd instruction should be a string');
        context.fillStyle = /** @type {string} */ (instruction[1]);
```


maybe I'm wrong somewhere?


If you could to send me a ready modified version of OL3 it would be very very nice for my work. 

Thank you again, your help is very precious for me.

Bye.

**Comment by MrPl0p** [comment]
Hi, 

I created a repo in my github account : https://github.com/MrPl0p/ol3-aswcr

The "ol-aswcr.js" is the js file which contains all overrided code. (aswcr = advanced styles with canvas renderer)
You just need to include it after the ol-debug.js file include.

I just tried it and its works : 
![image](https://cloud.githubusercontent.com/assets/11309450/8188109/a520b1ce-144b-11e5-8f4b-3b0c4c2ff0e1.png)

There's also a Readme file which explains how to.

If it's still not working, can you check the js error in the debug ?
I hope it will help you for your work.

**Comment by marcjansen** [comment]
Any chance this can be turned into a proper PR?

**Comment by effetx** [comment]
Great!!! Works perfectly!!!!
Many many many thanks!

In your opinion, there is a way to insert your code in ol3.js without using the debug version?

If I put your ol-aswcr.js below the ol.js Firebug tell me: goog is not defined.

Thank you again

**Comment by Morgul** [comment]
Looks like shortly (in the next month or two) I'm going to need this functionality myself. If @MrPl0p hasn't turned this into a pull request by then, I'll do the work to get this in.

**Comment by Jebjum** [comment]
Has there been any movement on this? Is MrPl0p's solution the best solution so far? Thanks.

**Comment by jeroenvheel** [comment]
works like a charm !!

**Comment by fredj** [comment]
fixed with #4632
